### ISSUE TITLE
Unable to use !== on union types with a string literal types

### ISSUE THREAD

**Issue by e98cuenc** [Microsoft/TypeScript#11277]
**TypeScript Version:**  2.0.3

**Code**

```ts
class Color {
	constructor(public r: number, public g: number, public b: number) {
	}
}

var f: 'none' | Color = new Color(0, 0, 0);
f !== 'none';
```

**Expected behavior:**
No errors

**Actual behavior:** 
error TS2365: Operator '!==' cannot be applied to types 'Color' and 'string'.

If the tested variable is a member of an object, the above code works:

```ts
class Color {
	constructor(public r: number, public g: number, public b: number) {}
}

var f: { x: 'none' | Color } = { x: new Color(0, 0, 0) }; 
f.x !== 'none';
```

**Comment by ahejlsberg** [comment]
This is working as intended. The error in the first example is reported because the control flow analyzer knows that the *actual* type of `f` in that location is `Color` (due to the assignment right above) and a `Color` value is not comparable to a `string` value (effectively the comparison doesn't make sense). Contrast that with this example

```ts
function foo(f: 'none' | Color) {
    f !== 'none';  // Ok
}
```

Here the comparison is ok because the control flow analyzer only knows the declared type for `f`, so `f` *might* be the string literal `'none'`. The same is true in your second example (although it arguable that we *could* know that `f.x` is a `Color`, we only track direct assignments to `f.x`).

If you really want a `'none'` value that you can always compare to a `'none' | Color`, you can explicitly give it that type in a type annotation:

```ts
const None: 'none' | Color = 'none';
```

Alternatively, you can use the expression `'none' as ('none' | Color)`.

**Comment by wallverb** [comment]
@ahejlsberg Isn't it the wrong assumption of control flow analyzer about actual type of `f` in that location ?

Consider:
```ts
class Color {
    constructor(public r: number, public g: number, public b: number) {
    }
}

var f: 'none' | Color = new Color(0, 0, 0);
function test() {
    f = 'none'
}
test()
f !== 'none'; 
```

**Comment by ahejlsberg** [comment]
@wallverb Effects of assignments in nested functions is an orthogonal issue. We currently don't account for them, and given the first class nature of functions we'll never be able to *completely* analyze such side effects without being overly conservative. There have been many discussions on this topic and #9998 summarizes where we are.

**Comment by wallverb** [comment]
Thank you @ahejlsberg . Great discussion

**Comment by aluanhaddad** [comment]
This feature is catching common logic errors and forcing people to write cleaner code that has fewer redundant conditionals. It's absolutely fantastic.
